compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEDmaCSE.cpp

```diff
@@ -0,0 +1,301 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file contains the a transformation pass that performs common
+// sub-expression elimination on AMDAIE DMA operations within the same region.
+// This eliminates identical circular DMA operations until only one exists. This
+// is different from the main upstream  CSE behaviour on ops with the `Pure`
+// trait in that `Pure` ops with no users will be eliminated, while DMA ops
+// without users are kept around.
+//
+// The implementation is based on and largely copied from:
+// https://github.com/llvm/llvm-project/blob/main/mlir/lib/Transforms/CSE.cpp
+//
+// TODO(jornt): It might be worthwile to try using the main upstream CSE pass,
```

Comment:
I spent some time trying to use upstream CSE instead (using Pure/MemoryEffects), but I wasn't able to get the same result. From what I am seeing, CSE would need changes to avoid accommodate this, but let me know if you see a way to achieve the same result with upstream CSE and/or a better way of accomplishing this.